### PR TITLE
feat: support Prisma 5

### PR DESCRIPTION
**Pull request by harelmo-lumigo**

### CODE REVIEW

File: src/hooks/prisma.test.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import { PrismaSpanBuilder } from '../../testUtils/prismaSpanBuilder';
+import { SpansContainer, TracerGlobals } from '../globals';
+import { getRandomId } from '../utils';
+import { hookPrisma } from './prisma';
+
+describe('Prisma', () => {
+  describe('bad Prisma versions', () => {
+    afterEach(() => {
+      // Reset overrides made to the prisma-client after require() in each test
+      jest.resetModules();
```
Comment (saartochner) [inline src/hooks/prisma.test.js:11]: this is actually really cool

File: src/hooks/prisma.test.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    test('produces simple operation span', async () => {
+      const count = await client.user.count({ where: { id: 123456 } });
+
+      const spans = SpansContainer.getSpans();
+      expect(spans).toHaveLength(1);
+
+      const expectedSpan = new PrismaSpanBuilder()
+        .withId(spans[0].id)
+        .withStarted(spans[0].started)
```
Comment (saartochner) [inline src/hooks/prisma.test.js:66]: can you explicitly check the this is actually filled? I.e. that `spans[0].started` is not undefined?
same for other fields

File: src/hooks/prisma.test.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+      const expectedRawQuerySpan = new PrismaSpanBuilder()
+        .withId(rawQuerySpan.id)
+        .withStarted(rawQuerySpan.started)
+        .withEnded(rawQuerySpan.ended)
+        .withOperation('$queryRaw')
+        .withQueryArgs(
+          `{"values":["${userId}"],"strings":["SELECT * FROM User WHERE id = ",";"]}`
+        )
+        .withResult(`[]`)
```
Comment (saartochner) [inline src/hooks/prisma.test.js:185]: can you add a test with a very long query / result? (i.e. `select * from users`)
we want to trim long contents

File: src/hooks/prisma.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  prismaLib.PrismaClient = function (...args) {
+    const originalInstance = new OriginalConstructor(...args);
+
+    if (typeof originalInstance.$extends !== 'function') {
+      return originalInstance;
+    }
+
+    return originalInstance.$extends({
```
Comment (saartochner) [inline src/hooks/prisma.ts:65]: do you want to return the response from the `$extends` function or `originalInstance`?
Can you add all your code (and specifically all the *added* code) to `safeExecute`?

File: src/hooks/prisma.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      query: {
```
Comment (saartochner) [inline src/hooks/prisma.ts:66]: Just to make sure that we're not missing something - what are other hooks that we can do (besides `query`)?

File: src/spans/awsSpan.ts [hunk 6]
```diff
@@ -50,6 +50,7 @@ export const MSSQL_SPAN = 'msSql';
 export const MYSQL_SPAN = 'mySql';
 export const NEO4J_SPAN = 'neo4j';
 export const ENRICHMENT_SPAN = 'enrichment';
+export const PRISMA_SPAN = 'prisma';
```
Comment (saartochner) [inline src/spans/awsSpan.ts:53]: Don't forget: I want to see it e2e in Lumigo. Let's make sure that you'll not get "span_type not found"

File: src/spans/common.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import { filterObjectKeys, getEventEntitySize } from '../utils';
+import { payloadStringify, truncate } from '../utils/payloadStringify';
+
+export const normalizeQuery = (query: string | any): string => {
+  if (typeof query === 'string') return truncate(query, getEventEntitySize());
```
[the remark this answers is not in the document]
Comment (saartochner) [reply src/spans/common.ts:5]: maybe I'm missing something, but I think that you don't use it in your code 😂

[the remark this answers is not in the document]
Comment (saartochner) [reply src/spans/common.ts:5]: please make sure to do it to the response as well

File: testUtils/prismaSetup.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { execSync } from 'child_process'
+
+export default () => {
+  // Clear previous test database and migrations used by Prisma
+  execSync('rm -rf ./testUtils/prisma/migrations && rm -rf ./testUtils/prisma/dev.db')
+  // Create a new database to be used by the Prisma tests
+  execSync('npx prisma migrate dev --name init --schema ./testUtils/prisma/schema.prisma')
```
Comment (saartochner) [inline testUtils/prismaSetup.js:7]: just to make sure:`npx` is something that is always installed with the runtime? is it additional requirement? Can you maybe do it in plain node (instead of `import { execSync } from 'child_process'`).
<not mandatory>

File: testUtils/prismaSpanBuilder.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      },
+      memoryAllocated: '1024',
+      messageVersion: 2,
+      parentId: HttpSpanBuilder.DEFAULT_PARENT_ID,
+      reporterAwsRequestId: HttpSpanBuilder.DEFAULT_PARENT_ID,
+      readiness: 'cold',
+      region: HttpSpanBuilder.DEFAULT_REGION,
+      invokedArn: HttpSpanBuilder.DEFAULT_ARN,
+      invokedVersion: HttpSpanBuilder.DEFAULT_VERSION,
+      runtime: 'AWS_Lambda_nodejs8.10',
```
Comment (saartochner) [inline testUtils/prismaSpanBuilder.js:25]: 👨‍🦳

File: testUtils/prismaSpanBuilder.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
+import { PRISMA_SPAN } from '../src/spans/awsSpan';
+import { HttpSpanBuilder } from './httpSpanBuilder';
+import { payloadStringify } from '@lumigo/node-core/lib/common';
+
+export class PrismaSpanBuilder {
+  constructor() {
+    this._span = {
```
Comment (saartochner) [inline testUtils/prismaSpanBuilder.js:7]: you copied a function_span, so it contains all kinds of fields that you real http_span will not have (e.g. readiness)

**Review by saartochner — CHANGES_REQUESTED** [review]
Very good job! 👏 
I would like to see the IT and the bottomline UI before the merge

File: src/hooks/prisma.ts (hunk 4 above)
Comment (harelmo-lumigo) [reply src/hooks/prisma.ts:65]: > do you want to return the response from the $extends function or originalInstance?

We have to return the output of `$extends`, as it returns a new client with the extended functionality and does not mutate `originalInstance` (actually, that's why I've added [this test](https://github.com/lumigo-io/lumigo-node/blob/ce130b9fdd9525538b8a87eecac9e720ef447aa9/src/hooks/prisma.test.js#L151)).

> Can you add all your code to safeExecute?

I guess you mean something like this:
```js
return safeExecute(
  () => originalInstance.$extends({
    query: {
      $allOperations: queryExtension,
    },
  }), 
  originalInstance
);
```
to return the original instance if `$extends` failed? yeah, it definitely makes sense

File: src/hooks/prisma.ts (hunk 5 above)
Comment (harelmo-lumigo) [reply src/hooks/prisma.ts:66]: All hooks are listed [here](https://www.prisma.io/docs/orm/prisma-client/client-extensions).
TL;DR - the other hooks are meant to _add functionality_ to the client / model level or mutate the query results, while `query` allows actually tapping into all operations

File: src/spans/awsSpan.ts (hunk 6 above)
Comment (harelmo-lumigo) [reply src/spans/awsSpan.ts:53]: You mean actual a test added to the _integration test_ repo with a Lambda running Prisma, or an E2E one-time "experiment" before we launch this?

File: src/spans/common.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (harelmo-lumigo) [reply src/spans/common.ts:5]: You're totally right, I've used it in an older iteration of the code but dropped it as it's irrelevant. Will revert the extraction to a shared module

File: testUtils/prismaSetup.js (hunk 8 above)
Comment (harelmo-lumigo) [reply testUtils/prismaSetup.js:7]: `npx` comes bundled with `npm` since npm@5, so I don't a reason why it won't work (we use npm@6, which I need to bump up actually - as the repo is working with Node 14 that Prisma does not support. Adam got us covered in his will [PR](https://github.com/lumigo-io/utils/pull/501/) 😂 )

File: src/spans/awsSpan.ts (hunk 6 above)
Comment (saartochner) [reply src/spans/awsSpan.ts:53]: The first is preferred, but if it gets rough then let go with the second :)

File: src/hooks/prisma.ts (hunk 4 above)
Comment (saartochner) [reply src/hooks/prisma.ts:65]: 👍

File: testUtils/prismaSetup.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import { execSync } from 'child_process'
+
+export default () => {
+  // Clear previous test database and migrations used by Prisma
+  execSync('rm -rf ./testUtils/prisma/migrations && rm -rf ./testUtils/prisma/dev.db')
```
[the remark this answers is not in the document]
Comment (harelmo-lumigo) [reply testUtils/prismaSetup.js]: Basically it's expanded to be relative to the process's CWD, which is the root of the project in that case, but I've change it to be more explicit [here](e436435af08b3e70eb3f5330d31aa4be746e3753):

File: testUtils/prismaSpanBuilder.js (hunk 9 above)
Comment (harelmo-lumigo) [reply testUtils/prismaSpanBuilder.js:25]: Just keepin' the tradition 👴

File: src/hooks/prisma.test.js (hunk 3 above)
Comment (harelmo-lumigo) [reply src/hooks/prisma.test.js:186]: Sure thing: 55742ae1027952ce285741cf3f65e3560d113876

File: testUtils/prismaSpanBuilder.js (hunk 10 above)
Comment (harelmo-lumigo) [reply testUtils/prismaSpanBuilder.js:7]: Got you... I copied `MongoSpanBuilder` which had all this :( I'll remove all the irrelevant fields

File: src/hooks/prisma.ts (hunk 4 above)
Comment (harelmo-lumigo) [reply src/hooks/prisma.ts]: 74c0301bea933dfd8e95c3e551f5a1bcb6f0046a

**Comment by harelmo-lumigo** [comment]
> Very good job! 👏 I would like to see the IT and the bottomline UI before the merge

Thanks :) 

Actually, I wanted to test it first in my personal dev-env. 
Do we have some kind of a formalized method to publish a "preview version" of the tracer to npm, or should I just install a locally-produced `tar.gz` on my lambda like a חורני and test it that way?

File: src/hooks/prisma.test.js (hunk 2 above)
Comment (saartochner) [reply src/hooks/prisma.test.js:79]: also ended 🙏

File: testUtils/prismaSpanBuilder.js (hunk 10 above)
Comment (harelmo-lumigo) [reply testUtils/prismaSpanBuilder.js:7]: Wait, I actually don't get it :) if the mongo-span has all those attributes too (e.g. `readiness`), is it by accident and those get ignored in the ingestion, or am I missing something here?

Comment (saartochner) [reply testUtils/prismaSpanBuilder.js:7]: are you sure that the mongo-span has these attributes in production as well, or is it there only in tests?
It makes no sense to have `readiness` on an HTTP span. So if it's there, then this is a bug (and then don't fix it now - just open a ticket :) )

Comment (harelmo-lumigo) [reply testUtils/prismaSpanBuilder.js:7]: The mongo-span uses [getBasicChildSpan](https://github.com/lumigo-io/lumigo-node/blob/2475be801a8db3eda04c23ce97b2125bff05a01b/src/spans/mongoDbSpan.ts#L12) which in turn uses [getBasicSpan](https://github.com/lumigo-io/lumigo-node/blob/2475be801a8db3eda04c23ce97b2125bff05a01b/src/spans/awsSpan.ts#L297), returning a bunch of attributes (such as [readiness](https://github.com/lumigo-io/lumigo-node/blob/2475be801a8db3eda04c23ce97b2125bff05a01b/src/spans/awsSpan.ts#L116)) that are probably pointless in that context, so you're definitely right :) 
I'll open a ticket for cleaning this up.

Comment (saartochner) [reply testUtils/prismaSpanBuilder.js:7]: awesome 💪

**Review by saartochner — APPROVED** [review]
